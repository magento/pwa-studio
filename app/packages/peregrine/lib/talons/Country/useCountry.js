import { useQuery } from '@apollo/client';

import DEFAULT_OPERATIONS from './country.gql';
import { mergeOperations } from '@magento/peregrine/lib/util/shallowMerge';

export const useCountry = () => {
    const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
    const { getCountriesQuery } = operations;

    const { data, error, loading } = useQuery(getCountriesQuery);

    let formattedCountriesData = [{ label: 'Loading Countries...', value: '' }];
    if (!loading && !error) {
        const { countries } = data;
        formattedCountriesData = countries.map(country => ({
            // If a country is missing the full english name just show the abbreviation.
            label: country.full_name_english || country.two_letter_abbreviation,
            value: country.two_letter_abbreviation
        }));
        formattedCountriesData.sort((a, b) => (a.label < b.label ? -1 : 1));
    }

    return {
        countries: formattedCountriesData,
        loading
    };
};
