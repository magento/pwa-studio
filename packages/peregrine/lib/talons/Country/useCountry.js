import { useQuery } from '@apollo/client';

export const useCountry = props => {
    const {
        queries: { getCountriesQuery }
    } = props;

    const { data, error, loading } = useQuery(getCountriesQuery);

    let formattedCountriesData = [{ label: 'Loading Countries...', value: '' }];
    if (!loading && !error) {
        const { countries } = data;
        formattedCountriesData = countries.map(country => ({
            // If a country is missing the full english name just show the abbreviation.
            label: country.full_name_english || country.two_letter_abbreviation,
            value: country.two_letter_abbreviation
        }));
        formattedCountriesData.sort((a, b) => (a.label < b.label ? -1 : 1));
    }

    return {
        countries: formattedCountriesData,
        loading
    };
};
