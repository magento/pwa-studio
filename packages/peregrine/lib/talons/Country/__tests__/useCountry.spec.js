import { useQuery } from '@apollo/client';

import { useCountry } from '../useCountry';

jest.mock('@apollo/client', () => ({
    useQuery: jest.fn().mockReturnValue({
        data: null,
        error: null,
        loading: true
    })
}));

const props = {
    queries: {}
};

test('returns loading placeholder option', () => {
    const talonProps = useCountry(props);
    expect(talonProps).toMatchSnapshot();
});

test('returns sorted data', () => {
    useQuery.mockReturnValueOnce({
        data: {
            countries: [
                {
                    two_letter_abbreviation: 'US',
                    full_name_english: 'United States'
                },
                {
                    two_letter_abbreviation: 'NL',
                    full_name_english: 'Netherlands'
                },
                {
                    two_letter_abbreviation: 'FR',
                    full_name_english: 'France'
                },
                // Sometimes full name is null, so we sort using
                // abbreviation.
                {
                    two_letter_abbreviation: 'AN',
                    full_name_english: null
                }
            ]
        },
        error: false,
        loading: false
    });

    const talonProps = useCountry(props);
    expect(talonProps).toMatchSnapshot();
});
