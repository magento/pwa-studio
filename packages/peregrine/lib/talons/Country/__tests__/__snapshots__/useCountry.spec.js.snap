// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns loading placeholder option 1`] = `
Object {
  "countries": Array [
    Object {
      "label": "Loading Countries...",
      "value": "",
    },
  ],
  "loading": true,
}
`;

exports[`returns sorted data 1`] = `
Object {
  "countries": Array [
    Object {
      "label": "AN",
      "value": "AN",
    },
    Object {
      "label": "France",
      "value": "FR",
    },
    Object {
      "label": "Netherlands",
      "value": "NL",
    },
    Object {
      "label": "United States",
      "value": "US",
    },
  ],
  "loading": false,
}
`;
